File: sdk/dart/lib/src/filter.dart
import 'dart:convert';

class Filter {
  final Map<String, dynamic> _node;

  Filter._(this._node);

  static Filter eq(String field, dynamic value) {
    return Filter._({'type': 'eq', 'field': field, 'value': value});
  }

  static Filter neq(String field, dynamic value) {
    return Filter._({'type': 'neq', 'field': field, 'value': value});
  }

  static Filter gt(String field, dynamic value) {
    return Filter._({'type': 'gt', 'field': field, 'value': value});
  }

  static Filter and(List<Filter> filters) {
    return Filter._({'type': 'and', 'value': filters.map((f) => f._node).toList()});
  }

  static Filter or(List<Filter> filters) {
    return Filter._({'type': 'or', 'value': filters.map((f) => f._node).toList()});
  }

  String toJsonString() {
    return jsonEncode(_node);
  }
}

File: sdk/php/src/Filter.php
<?php
namespace LibraVDB;

class Filter {
    public static function eq(string $field, $value): array {
        return ["type" => "eq", "field" => $field, "value" => $value];
    }

    public static function gt(string $field, $value): array {
        return ["type" => "gt", "field" => $field, "value" => $value];
    }

    public static function lt(string $field, $value): array {
        return ["type" => "lt", "field" => $field, "value" => $value];
    }

    public static function in(string $field, array $values): array {
        return ["type" => "contains_any", "field" => $field, "values" => $values];
    }

    public static function and(array ...$filters): array {
        return ["type" => "and", "filters" => $filters];
    }

    public static function or(array ...$filters): array {
        return ["type" => "or", "filters" => $filters];
    }

    public static function toJson(array $filter): string {
        return json_encode($filter);
    }
}
